File: 547-number-of-provinces/number-of-provinces.java
class Solution {

    public void dfs(Map<Integer, List<Integer>> adj, int u, boolean[] visited) {
        visited[u] = true;

        for (int v : adj.get(u)) {
            if (!visited[v]) {
                dfs(adj, v, visited);
            }
        }
    }
    public int findCircleNum(int[][] isConnected) {
        HashMap<Integer, List<Integer>> mp= new HashMap();

        int n= isConnected.length;
        int m= isConnected[0].length;

      for (int i = 0; i < n; i++) {
            mp.put(i, new ArrayList<>());
            for (int j = 0; j < n; j++) {
                if (isConnected[i][j] == 1 && i != j) {
                    mp.get(i).add(j);
                }
            }
        }

        boolean[] visited = new boolean[n];
        int count = 0;

        // DFS on all components
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                count++;
                dfs(mp, i, visited);
            }
        }

        return count;
    }
}

File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    void dfs(int node, vector<bool> &visited, unordered_map<int, vector<int>> &adj){
        if(visited[node]){
            return ;
        }

        visited[node]=true;
        
        for(int &ng: adj[node]){
            if(!visited[ng]){
                dfs(ng, visited, adj);
            }
        }

    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n= isConnected.size();

        unordered_map<int, vector<int>> adj;

        for(int i=0; i<n; i++){
            for(int j=0; j< n; j++){
                if(isConnected[i][j]==1){
                    adj[i].push_back(j);
                    adj[j].push_back(i);
                }
            }
        }

        vector<bool> visited(n, false);
        int count=0;

        for(int i=0;i< n; i++){
            if(!visited[i]){
                count++;
                dfs(i, visited, adj);
            }
        }

        return count;
    }
};